File: model_1/CS3233_2024_Midterm/hero_leveling_1_solution.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <climits>
using namespace std;

typedef long long ll;

const int MAX_SUM = 23000;
const ll INF = 1.5e18;

int main() {
    int n, k;
    ll L;
    cin >> n >> k >> L;
    vector<tuple<ll, ll, int>> quests(n);
    for (int i = 0; i < n; i++) {
        ll M, X;
        int T;
        cin >> M >> X >> T;
        quests[i] = make_tuple(M, X, T);
    }
    sort(quests.begin(), quests.end());

    vector<vector<ll>> dp(n + 1, vector<ll>(MAX_SUM + 1, -INF));
    dp[0][0] = 0;

    for (int idx = 0; idx < n; idx++) {
        auto [M, X, T] = quests[idx];
        auto new_dp = dp;
        for (int j = n - 1; j >= 0; j--) {
            for (int s = 0; s <= MAX_SUM; s++) {
                if (dp[j][s] == -INF) continue;
                if (dp[j][s] < M) continue;
                ll new_level = dp[j][s] + X;
                if (new_level > INF) new_level = INF;
                int next_j = j + 1;
                int next_s = s;
                if (next_j > k) {
                    if (s + T > MAX_SUM) continue;
                    next_s = s + T;
                }
                if (new_level > new_dp[next_j][next_s]) {
                    new_dp[next_j][next_s] = new_level;
                }
            }
        }
        dp = new_dp;
    }

    ll ans = INF;
    for (int j = 1; j <= n; j++) {
        for (int s = 0; s <= MAX_SUM; s++) {
            if (dp[j][s] >= L) {
                if (j <= k) {
                    int total_trauma = 0;
                    for (int i = 0; i < j; i++) {
                        auto [M, X, T] = quests[i];
                        total_trauma += T;
                    }
                    if (total_trauma < ans) {
                        ans = total_trauma;
                    }
                } else {
                    ll trauma_rating = total_trauma - s;
                    if (trauma_rating < ans) {
                        ans = trauma_rating;
                    }
                }
            }
        }
    }
    cout << ans << endl;
    return 0;
}

File: model_1/CS3233_2024_Midterm/hero_leveling_1_solution.py
import sys

def main():
    data = sys.stdin.read().split()
    if not data:
        return
    it = iter(data)
    n = int(next(it))
    k_val = int(next(it))
    L_val = int(next(it))
    quests = []
    for _ in range(n):
        M = int(next(it))
        X = int(next(it))
        T = int(next(it))
        quests.append((M, X, T))
    
    quests.sort(key=lambda x: (x[0], x[1], x[2]))
    
    MAX_TRAUMA_SUM = 22600
    dp = [[-10**18] * (MAX_TRAUMA_SUM + 1) for _ in range(n + 1)]
    dp[0][0] = 0
    
    for quest in quests:
        m, x, t = quest
        new_dp = [row[:] for row in dp]
        for j in range(n - 1, -1, -1):
            for s_val in range(MAX_TRAUMA_SUM + 1):
                if dp[j][s_val] < m:
                    continue
                new_level = dp[j][s_val] + x
                if new_level > 10**18:
                    new_level = 10**18
                if j < k_val:
                    ns = s_val + t
                    if ns <= MAX_TRAUMA_SUM:
                        if new_level > new_dp[j + 1][ns]:
                            new_dp[j + 1][ns] = new_level
                else:
                    ns = s_val + t
                    if ns <= MAX_TRAUMA_SUM:
                        if new_level > new_dp[j + 1][ns]:
                            new_dp[j + 1][ns] = new_level
        dp = new_dp

    ans = 10**18
    for j in range(1, n + 1):
        for s_val in range(MAX_TRAUMA_SUM + 1):
            if dp[j][s_val] >= L_val:
                if j <= k_val:
                    total_trauma = s_val
                    if total_trauma < ans:
                        ans = total_trauma
                else:
                    total_trauma = s_val
                    if total_trauma < ans:
                        ans = total_trauma
    print(ans)

if __name__ == '__main__':
    main()